models/ResourceService.cfc

```diff
@@ -228,33 +228,41 @@ component singleton accessors="true" {
-		var resourceBundle = {};
-		var thisKEY        = "";
-		var rbFilePath     = arguments.rbFile & iIf(
-			len( arguments.rbLocale ),
-			de( "_" ),
-			de( "" )
-		) & arguments.rbLocale & ".properties";
-
-		var rbFilePath = ( len( arguments.rbLocale ) ) ? "#arguments.rBFile#_#arguments.rbLocale#.properties" : "#arguments.rBFile#.properties";
-
-		// read file
-		var fis = getResourceFileInputStream( rbFilePath );
-		var fir = createObject( "java", "java.io.InputStreamReader" ).init( fis, "UTF-8" );
-
-		// Init RB with file Stream
-		var rb = createObject( "java", "java.util.PropertyResourceBundle" ).init( fir );
-		try {
-			// Get Keys
-			var keys = rb.getKeys();
-
-			// Loop through property keys and store the values into bundle
-			while ( keys.hasMoreElements() ) {
-				thisKEY                   = keys.nextElement();
-				resourceBundle[ thisKEY ] = rb.handleGetObject( thisKEY );
-			}
-		} finally {
-			fis.close();
+		// try to load hierarchical resource rbfile_LANG_COUNTRY_VARIANT, start with default resource, followed by 
+		// rbFile.properties, rbFile_LANG.properties, rbFile_LANG_COUNTRY.properties, rbFile_LANG_COUNTRY_VARIANT.properties (assuming LANG, COUNTRY and VARIANT are present)
+		// All items in resourceBundle will be overwritten by more specific ones.			
+
+		// Create all file options from locale
+		var myRbFile = arguments.rbFile;
+		//start with default locale
+		var smartBundleFiles = ["#myRbFile#_#variables.defaultLocale#.properties"];
+		//add base resource, without language, country or variant
+		smartBundleFiles.append("#myRbFile#.properties");
+		//include lang, country and variant (if present)
+		// extract and add to bundleArray by splitting rbLocale as list on '_'
+		arguments.rbLocale.listEach( function( localePart, index, list){
+			myRbFile &= "_#localePart#";
+			smartBundleFiles.append("#myRbFile#.properties");	
+		}, "_" );
+		
+		// load all resource files for all default,lang, country and variants 
+		// and overwrite parent keys when present so you you will always have defaults
+		// AND specific resource values for countries and variants without duplicating everything.
+		var resourceBundle = structNew();
+		var IsValidBundleLoaded = false;
+		smartBundleFiles.each(function(resourceFile){
+			var ResourceBundleFullPath = variables.controller.locateFilePath( resourceFile );
+			if ( ResourceBundleFullPath.len() ) {
+				resourceBundle.append( _loadSubBundle( ResourceBundleFullPath ), true ); //append and overwrite
+				IsValidBundleLoaded = true; // at least one bundle loaded so no errors
+			};
+		});
+
+		// Validate resource is loaded or error.
+		if( !IsValidBundleLoaded ){
+			var rbFilePath = "#arguments.rbFile#_#arguments.rbLocale#.properties";
+			var rbFullPath = variables.controller.locateFilePath( rbFilePath );
+			throw("The resource bundle file: #rbFilePath# does not exist. Please check your path", "FullPath: #rbFullPath#", "ResourceBundle.InvalidBundlePath");
```

Comment:
Use named arguments please, better, for visibility: `message, type, details`

Reply:
Also, the method annotations should have a `@throws InvalidBundlePath`

Reply:
Ok fixed named arguments for throw in more places and added @throws in more places (getResourceFileInputStream can throw this)